sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCast.scala

```diff
@@ -0,0 +1,155 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.Literal.FalseLiteral
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+
+/**
+ * Unwrap casts in binary comparison operations with the following pattern:
+ *   `BinaryComparison(Cast(fromExp, _), Literal(value, toType))`
+ * This rule optimize expressions with this pattern by either replacing the cast with simpler
+ * constructs, or moving the cast from the expression side to the literal side, so they can be
+ * optimized away and pushed down to data sources.
+ *
+ * Currently this only handles the case where `fromType` (of `fromExp`) and `toType` are of
+ * integral types (i.e., byte, short, int and long). It checks to see if the literal `value` is
+ * within range (min, max) of the `fromType`. If this is true then it means we can safely cast the
+ * `value` to the `fromType` and thus able to move the cast to the literal side. Otherwise, it
+ * replaces the cast with different simpler constructs, such as
+ * `EqualTo(fromExp, Literal(max, fromType)` when input is
+ * `GreaterThanOrEqualTo(Cast(fromExp, fromType), Literal(max, toType))`
+ */
+object UnwrapCast extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l: LogicalPlan => l transformExpressionsUp {
+      case e @ BinaryComparison(_, _) => unwrapCast(e)
+    }
+  }
+
+  private def unwrapCast(exp: Expression): Expression = exp match {
+    case BinaryComparison(Literal(_, _), Cast(_, _, _)) =>
+      def swap(e: Expression): Expression = e match {
+        case GreaterThan(left, right) => LessThan(right, left)
+        case GreaterThanOrEqual(left, right) => LessThanOrEqual(right, left)
+        case EqualTo(left, right) => EqualTo(right, left)
+        case EqualNullSafe(left, right) => EqualNullSafe(right, left)
+        case LessThanOrEqual(left, right) => GreaterThanOrEqual(right, left)
+        case LessThan(left, right) => GreaterThan(right, left)
+        case other => other
+      }
+      swap(unwrapCast(swap(exp)))
+
+    case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType)) =>
```

Comment:
```
case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType: IntegralType)) =>
```
You can do the above that will ensure `toType` will be `IntergralType`. And you don't need to do asInstanceOf laster if you have this.

Reply:
👍 yup will do this

Reply:
actually since we plan to handle more types later, this will be replaced. I'll probably do:
```scala
case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType)) if canImplicitlyCast(fromExp, toType) =>
```
where `canImplicitlyCast` will encapsulate the logic on whether the case is valid.